pkg/controllers/jobset_controller.go

```diff
@@ -30,30 +35,284 @@ type JobSetReconciler struct {
 		Complete(r)
 }
+
+func (r *JobSetReconciler) constructJobFromTemplate(js *jobset.JobSet, jobTemplate *jobset.ReplicatedJob) (*batchv1.Job, error) {
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      make(map[string]string),
+			Annotations: make(map[string]string),
+			Name:        genJobName(js, jobTemplate),
+			Namespace:   js.Namespace,
+		},
+		Spec: *jobTemplate.Template.Spec.DeepCopy(),
+	}
+
+	// If enableDNSHostnames is set, update job spec to set subdomain as
+	// job name (a headless service with same name as job will be created later).
+	if jobTemplate.Network.EnableDNSHostnames != nil && *jobTemplate.Network.EnableDNSHostnames {
+		job.Spec.Template.Spec.Subdomain = job.Name
+	}
+
+	// Set controller owner reference for garbage collection and reconcilation.
+	if err := ctrl.SetControllerReference(js, job, r.Scheme); err != nil {
+		return nil, err
+	}
+	return job, nil
+}
+
+// getChildJobs fetches all Jobs owned by the JobSet and returns them
+// categorized by status (active, successful, failed).
+func (r *JobSetReconciler) getChildJobs(ctx context.Context, js *jobset.JobSet, req ctrl.Request) (*childJobs, error) {
+	// Get all active jobs owned by JobSet.
+	var childJobList batchv1.JobList
+	if err := r.List(ctx, &childJobList, client.InNamespace(req.Namespace), client.MatchingFields{jobOwnerKey: req.Name}); err != nil {
+		klog.Error(err, "unable to list child Jobs")
+		return nil, err
+	}
+
+	jobs := childJobs{}
+	for i, job := range childJobList.Items {
+		_, finishedType := isJobFinished(&job)
+		switch finishedType {
+		case "": // active
+			jobs.active = append(jobs.active, &childJobList.Items[i])
+		case batchv1.JobFailed:
+			jobs.failed = append(jobs.failed, &childJobList.Items[i])
+		case batchv1.JobComplete:
+			jobs.successful = append(jobs.successful, &childJobList.Items[i])
+		}
+	}
+
+	return &jobs, nil
+}
+
+func (r *JobSetReconciler) createJobs(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+	for _, jobTemplate := range js.Spec.Jobs {
+		job, err := r.constructJobFromTemplate(js, &jobTemplate)
+		if err != nil {
+			klog.Errorf("error constructing job from template %s: %v", jobTemplate.Name, err)
+			return err
+		}
+
+		// Check if we need to create this job.
+		// If not, skip this job and continue iterating to the next job.
+		if create := r.shouldCreateJob(ctx, job, ownedJobs); !create {
+			klog.Infof("job %s already exists, not creating", job.Name)
+			continue
+		}
+
+		// Create headless service if specified for this job.
+		if jobTemplate.Network.EnableDNSHostnames != nil && *jobTemplate.Network.EnableDNSHostnames {
+			klog.Infof("creating headless service: %s", job.Name)
+			if err := r.createHeadlessSvcIfNotExist(ctx, js, job); err != nil {
+				klog.Errorf("error creating headless service: %v", err)
+				return err
+			}
+		}
+
+		// Create the job.
+		klog.Infof("creating job %s", job.Name)
+		if err := r.Create(ctx, job); err != nil {
```

Comment:
We need to think about how to deal with the case where the job exists but not owned by jobset

Reply:
Added TODO for this